lib/auth.js

```diff
@@ -0,0 +1,141 @@
+      }
+    } else {
+      if (typeof execution[field] === 'undefined') {
+        execution[field] = null
+      } else if (execution[field] !== null) {
+        this[kSetForbiddenFieldsToNull](execution[field], path.slice(1))
+      }
+    }
+  }
+
+  async [kApplyPolicyToSelections] (fieldTypeMap, parentDefinition, definition, context, cursor) {
+    const selectionsToRemove = []
+    const errors = []
+    const nestedSelectionsPromises = []
+    await Promise.all(definition.selectionSet.selections.map(async selection => {
+      // Build cursor so we know where we are
+      const newCursor = this[kBuildCursor](cursor, selection)
+
+      // Check if protected field and get field auth directive AST
+      const field = fieldTypeMap[selection.name.value]
+      const fieldAuthDirectiveAST = this[kGetFieldAuthDirectiveAST](field)
+
+      // If protected field, apply policy and apply errors
+      if (fieldAuthDirectiveAST !== null && !await this[kApplyPolicy](fieldAuthDirectiveAST, context, field)) {
+        errors.push(new GraphQLError('auth error', selection, undefined, undefined, newCursor.path))
+        selectionsToRemove.push(selection.name.value)
+        return
+      }
+
+      // Continue down the tree if we have further fields to check
+      const namedType = getNamedType(field.type)
+      if (typeof namedType.getFields === 'function') {
+        nestedSelectionsPromises.push(this[kApplyPolicyToSelections](namedType.getFields(), definition, selection, context, newCursor))
+      }
+    }))
+
+    // Apply policy to nested selections
+    const nestedErrors = await Promise.all(nestedSelectionsPromises)
+    errors.push(...nestedErrors.flat())
+
+    // Remove selections that failed the policy so they are not executed
+    if (selectionsToRemove.length > 0) {
+      if (parentDefinition !== null && selectionsToRemove.length === definition.selectionSet.selections.length) {
+        parentDefinition.selectionSet.selections = parentDefinition.selectionSet.selections.filter(selection => selection.name.value !== definition.name.value)
+      } else {
+        definition.selectionSet.selections = definition.selectionSet.selections.filter(selection => !selectionsToRemove.includes(selection.name.value))
+      }
+    }
+    return errors
+  }
+
+  async authContextHook (_schema, _document, context) {
+    const auth = await this[kAuthContext](context)
+    Object.assign(context, { auth })
+  }
+
+  async applyPolicyHook (schema, document, context) {
+    context[kAuthErrors] = null
+    const authErrors = []
+
+    // Apply policy to document definitions
+    await Promise.all(document.definitions.map(async definition => {
+      const schemaTypeMap = schema.getType(upperFirst(definition.operation)).getFields()
+
+      const errors = await this[kApplyPolicyToSelections](schemaTypeMap, null, definition, context, { path: [] })
+      authErrors.push(...errors)
+    }))
```

Comment:
I think this is not strictly correct as I think we need to apply these to the real objects instead. Maybe an approach that overrides the resolvers?

Reply:
Yeah that could work pretty nicely for a couple of reasons!
 - more performant - no need to traverse the query AST
 - easy access to all the parent objects etc straight away - solving: https://github.com/mercurius-js/auth/pull/3#discussion_r615383879

My only concern would be in gateway mode. I.e. when the federated schema changes, the resolvers wouldn't necessarily reflect the latest schema. I'm thinking of solving this by maybe adding a mercurius application lifecycle hook that triggers when the federated schema changes so we can re-override the resolvers?

Reply:
agreed!

Reply:
Benchmarks are looking a lot better after refactoring to use resolver overrides! Still got a fair bit to polish up (including a PR in mercurius to make), but looking very promising!

```
==============================
= Normal Mode | Without Auth =
==============================
Running 10s test @ http://localhost:3000/graphql
100 connections
50 workers

┌─────────┬──────┬──────┬───────┬───────┬─────────┬────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev  │ Max    │
├─────────┼──────┼──────┼───────┼───────┼─────────┼────────┼────────┤
│ Latency │ 0 ms │ 8 ms │ 18 ms │ 23 ms │ 8.38 ms │ 4.5 ms │ 129 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%  │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 5391    │ 5391    │ 10679   │ 11023  │ 10137   │ 1624.79 │ 5391    │
├───────────┼─────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.84 MB │ 2.84 MB │ 5.62 MB │ 5.8 MB │ 5.33 MB │ 854 kB  │ 2.84 MB │
└───────────┴─────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

113k requests in 10.02s, 59.4 MB read
===========================
= Normal Mode | With Auth =
===========================
Running 10s test @ http://localhost:3000/graphql
100 connections
50 workers

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max    │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼────────┤
│ Latency │ 0 ms │ 9 ms │ 19 ms │ 24 ms │ 9.16 ms │ 4.42 ms │ 171 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 5083    │ 5083    │ 9903    │ 10015   │ 9407.8  │ 1447.56 │ 5082    │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.67 MB │ 2.67 MB │ 5.21 MB │ 5.27 MB │ 4.95 MB │ 761 kB  │ 2.67 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

104k requests in 10.02s, 54.8 MB read
```